server/src/main/java/org/elasticsearch/cluster/service/ClusterApplierService.java

```diff
@@ -216,7 +216,24 @@ public void removeApplier(ClusterStateApplier applier) {
      * Add a listener for updated cluster states
      */
     public void addListener(ClusterStateListener listener) {
-        clusterStateListeners.add(listener);
+        clusterStateListeners.add(new SafeClusterStateListener(listener, threadPool.getThreadContext().newRestorableContext(true)));
+    }
+
+    private static class SafeClusterStateListener implements ClusterStateListener {
```

Comment:
Let's call this `ContextPreservingClusterStateListener` - I think the ones in the `MasterService` were called `Safe` because they caught all exceptions, and they're renamed/gone in 5f21df692f60a9b542ae71f1129c290f27f7775a anyway.